examples/ct/structures.go

```diff
@@ -99,15 +112,208 @@ func SignV1SCTForPrecertificate(km crypto.KeyManager, cert *x509.Certificate, t
+// well make this available.
+func WriteTimestampedEntry(w io.Writer, t ct.TimestampedEntry) error {
+	if err := binary.Write(w, binary.BigEndian, &t.Timestamp); err != nil {
+		return err
+	}
+	if err := binary.Write(w, binary.BigEndian, &t.EntryType); err != nil {
+		return err
+	}
+	switch t.EntryType {
+	case ct.X509LogEntryType:
+		if err := writeVarBytes(w, t.X509Entry, ct.CertificateLengthBytes); err != nil {
+			return err
+		}
+	case ct.PrecertLogEntryType:
+		if err := binary.Write(w, binary.BigEndian, t.PrecertEntry.IssuerKeyHash); err != nil {
+			return err
+		}
+		if err := writeVarBytes(w, t.PrecertEntry.TBSCertificate, ct.PreCertificateLengthBytes); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown EntryType: %d", t.EntryType)
+	}
+
+	return writeVarBytes(w, t.Extensions, ct.ExtensionsLengthBytes)
+}
+
+// WriteMerkleTreeLeaf writes a MerkleTreeLeaf in the binary format specified by RFC 6962.
+// The CT go code includes a deserializer but not a serializer and we might as well make this
+// available to other users.
+func WriteMerkleTreeLeaf(w io.Writer, l ct.MerkleTreeLeaf) error {
+	if l.Version != ct.V1 {
+		return fmt.Errorf("unknown Version: %d", l.Version)
+	}
+
+	if l.LeafType != ct.TimestampedEntryLeafType {
+		return fmt.Errorf("unknown LeafType: %d", l.LeafType)
+	}
+
+	if err := binary.Write(w, binary.BigEndian, l.Version); err != nil {
+		return err
+	}
+	if err := binary.Write(w, binary.BigEndian, l.LeafType); err != nil {
+		return err
+	}
+	if err := WriteTimestampedEntry(w, l.TimestampedEntry); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Serialize writes out a CTLogEntry in binary form. This is not an RFC 6962 data structure
+// and is only used internally by the log.
+func (c CTLogEntry) Serialize(w io.Writer) error {
+	if err := WriteMerkleTreeLeaf(w, c.Leaf); err != nil {
+		return err
+	}
+
+	if err := writeUint(w, uint64(len(c.Chain)), 2); err != nil {
+		return err
+	}
+
+	for _, certBytes := range c.Chain {
+		if err := writeVarBytes(w, certBytes, 4); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Deserialize reads a binary format CTLogEntry and stores the result into an existing
+// CTLogEntry struct. This is an internal data structure and is not defined in RFC 6962 or
+// exposed to clients.
+func (c *CTLogEntry) Deserialize(r io.Reader) error {
+	leaf, err := ct.ReadMerkleTreeLeaf(r)
+
+	if err != nil {
+		return err
+	}
+
+	c.Leaf = *leaf
+
+	numCerts, err := readUint(r, 2)
+
+	if err != nil {
+		return err
+	}
+
+	chain := make([]ct.ASN1Cert, 0, numCerts)
+	var cert uint64
+
+	for cert = 0; cert < numCerts; cert++ {
+		certBytes, err := readVarBytes(r, 4)
+
+		if err != nil {
+			return err
+		}
+
+		chain = append(chain, certBytes)
+	}
+
+	c.Chain = chain
+
+	return nil
+}
+
+// These came from the CT go code. Currently don't want to push changes upstream to make
+// them visible but this is an option for the future.
```

Comment:
Yes, that's part of the argument for doing it. I might do it and refactor this once I've merged it and got the add-chain handler working. I need this to finish the handler.